### PR TITLE
Prevent functorch transforms from inheriting Dynamo metadata

### LINKED ISSUE #154036: torch.compile on torch.vmap function gives different shape to torch.vmap alone when using jacrev
### 🐛 Describe the bug

When using `torch.compile` on a function that is vmapped and contains a call to `jacrev` or `jacfwd`, the resulting output shape and values differ from the uncompiled version. A minimal reproducer is as follows:

```python
import torch

burgers_viscosity = 0.07

def phi_fn(x, time):
    denominator = 4*burgers_viscosity*(time+1)
    return torch.exp(-torch.square(x - 4*time)/denominator) + torch.exp(-torch.square(x - 4*time - 2*torch.pi)/denominator)

def fn(x, time):
    grad_res, res = torch.func.jacrev(lambda x: tuple([phi_fn(x, time)]*2), has_aux=True)(x+torch.pi)
    return -(2*burgers_viscosity/res)*grad_res + 4


compile_options = dict(
    mode=None, backend="inductor", fullgraph=True
)

@torch.compile(**compile_options)
def phi_fn_compiled(x, time):
    denominator = 4*burgers_viscosity*(time+1)
    return torch.exp(-torch.square(x - 4*time)/denominator) + torch.exp(-torch.square(x - 4*time - 2*torch.pi)/denominator)

@torch.compile(**compile_options)
def fn_compiled(x, time):
    grad_res, res = torch.func.jacrev(lambda x: tuple([phi_fn_compiled(x, time)]*2), has_aux=True)(x+torch.pi)
    return -(2*burgers_viscosity/res)*grad_res + 4

xvals = torch.linspace(-torch.pi, torch.pi, 128, dtype=torch.float64)

vmapped_phi_fn = torch.vmap(phi_fn, in_dims=(0, None))
vmapped_fn = torch.vmap(fn, in_dims=(0, None))

vmapped_compiled_phi_fn = torch.compile(torch.vmap(phi_fn_compiled, in_dims=(0, None)), **compile_options)
vmapped_compiled_fn = torch.compile(torch.vmap(fn_compiled, in_dims=(0, None)), **compile_options)

print("phi_fn() == phi_fn_compiled():", torch.allclose(phi_fn(xvals, 0.0), phi_fn_compiled(xvals, 0.0)))
print("\tshape(phi_fn()) =", phi_fn(xvals, 0.0).shape, "\n\tshape(phi_fn_compiled()) =", phi_fn_compiled(xvals, 0.0).shape)

print()
print("fn() == fn_compiled():", torch.allclose(fn(xvals, 0.0), fn_compiled(xvals, 0.0)))
print("\tshape(fn()) =", fn(xvals, 0.0).shape, "\n\tshape(fn_compiled()) =", fn_compiled(xvals, 0.0).shape)

print()
print("vmapped_phi_fn() == vmapped_compiled_phi_fn():", torch.allclose(vmapped_phi_fn(xvals, 0.0), vmapped_compiled_phi_fn(xvals, 0.0)))
print("\tshape(vmapped_phi_fn()) =", vmapped_phi_fn(xvals, 0.0).shape, "\n\tshape(vmapped_compiled_phi_fn()) =", vmapped_compiled_phi_fn(xvals, 0.0).shape)

print()
print("vmapped_fn == vmapped_compiled_fn:", torch.allclose(vmapped_fn(xvals, 0.0), vmapped_compiled_fn(xvals, 0.0)))
print("\tshape(vmapped_fn()) =", vmapped_fn(xvals, 0.0).shape, "\n\tshape(vmapped_compiled_fn()) =", vmapped_compiled_fn(xvals, 0.0).shape)
```

The output of which is:
```python
phi_fn() == phi_fn_compiled(): True
	shape(phi_fn()) = torch.Size([128]) 
	shape(phi_fn_compiled()) = torch.Size([128])

fn() == fn_compiled(): True
	shape(fn()) = torch.Size([128, 128]) 
	shape(fn_compiled()) = torch.Size([128, 128])

vmapped_phi_fn() == vmapped_compiled_phi_fn(): True
	shape(vmapped_phi_fn()) = torch.Size([128]) 
	shape(vmapped_compiled_phi_fn()) = torch.Size([128])

vmapped_fn == vmapped_compiled_fn: False
	shape(vmapped_fn()) = torch.Size([128]) 
	shape(vmapped_compiled_fn()) = torch.Size([128, 128])
```

### Error logs

_No response_

### Versions

Collecting environment information...
PyTorch version: 2.7.0+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 11 (bullseye) (x86_64)
GCC version: (Debian 10.2.1-6) 10.2.1 20210110
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.12.10 (main, Apr  9 2025, 00:30:25) [GCC 10.2.1 20210110] (64-bit runtime)
Python platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3070 Laptop GPU
Nvidia driver version: 576.28
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
Address sizes:                        39 bits physical, 48 bits virtual
CPU(s):                               16
On-line CPU(s) list:                  0-15
Thread(s) per core:                   2
Core(s) per socket:                   8
Socket(s):                            1
Vendor ID:                            GenuineIntel
CPU family:                           6
Model:                                141
Model name:                           11th Gen Intel(R) Core(TM) i7-11800H @ 2.30GHz
Stepping:                             1
CPU MHz:                              2303.999
BogoMIPS:                             4607.99
Virtualization:                       VT-x
Hypervisor vendor:                    Microsoft
Virtualization type:                  full
L1d cache:                            384 KiB
L1i cache:                            256 KiB
L2 cache:                             10 MiB
L3 cache:                             24 MiB
Vulnerability Gather data sampling:   Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves avx512vbmi umip avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg avx512_vpopcntdq rdpid movdiri movdir64b fsrm avx512_vp2intersect md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] Could not collect
[conda] Could not collect

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185547

Functorch transforms preserve Python metadata with functools.wraps. That also copies the wrapped callable's __dict__, so transforming a torch.compile result could leave Dynamo compile-wrapper metadata on the new transform wrapper. A later torch.compile on that wrapper could then see stale attributes such as get_compiler_config, _torchdynamo_inline, or _is_torch_compile from the inner compiled callable instead of metadata for the transformed function. For the reported vmap(jacrev(...)) case, that allowed Dynamo to confuse the vmapped wrapper with the original compiled function and either assert while installing compiler config or lose the vmap behavior, producing an extra jacrev dimension.

Fix this at the transform boundary by keeping normal functools.wraps metadata but scrubbing the Dynamo wrapper attributes that describe the original compiled callable. This now covers vmap/chunk_vmap and the public grad, grad_and_value, jacrev, jacfwd, hessian, and functionalize paths. I kept the fix local to functorch wrappers instead of changing Dynamo wrapper semantics or dropping functools.wraps entirely, because callers still expect normal function metadata to be preserved and only the copied compile state is invalid.

Fixes #154036

Generated by my agent

Test Plan:

python test/dynamo/test_higher_order_ops.py -k test_compile_vmap_compiled_fn_jacrev_has_aux

python test/dynamo/test_higher_order_ops.py -k test_compile_function_transform_wrappers_compiled_fn

python test/dynamo/test_higher_order_ops.py -k test_functionalize_compiled_fn_does_not_copy_dynamo_attrs

lintrunner -a

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and expanded the metadata scrubber to cover grad, grad_and_value, jacrev, jacfwd, hessian, and functionalize in addition to vmap/chunk_vmap. I also added tests that assert all known Dynamo compile-wrapper attributes, including _is_torch_compile, are not copied before recompiling transformed functions.

**Comment by jansel** [comment]
My agent says I rebased the PR onto current main and resubmitted it to refresh CI after the docs preview jobs failed at the upload step with aws: command not found. The focused Dynamo/functorch tests still pass; full lint is currently blocked by an unrelated current-main Pyrefly error in torch/nn/modules/loss.py, while lint on the PR diff with Pyrefly skipped passes.

**Review by zou3519 — APPROVED** [review]
